### DISCUSSION TITLE
Suggestions of @gkiar

### DISCUSSION THREAD

**Discussion by yohanchatelain** [yohanchatelain/sigdigits#3] [General]
Repost @gkiar comments:
"
This looks great! A few questions that may either be a) entirely irrelevant, or b) a tiny bit helpful  .
 
https://github.com/yohanchatelain/sigdigits/blob/2b4de1d895bab8db8f50ee5ff7979a03109e3941/sigdigits.py#L84-L85
 
* Q: Do we want to fix a single random seed if we're shuffling, or fix and then compute the value over a few iterations and average them? I don't think we want a measure that changes with random state, if we can avoid it. This question applies a few more times throughout the module.
 
https://github.com/yohanchatelain/sigdigits/blob/2b4de1d895bab8db8f50ee5ff7979a03109e3941/sigdigits.py#L142-L149
 
* Q: if all we're doing here is comparing the `z` value to the `2^{-k}` power for each iteration, couldn't we just take the `log` of all our values of `z` (that are non-zero) and skip the loop all together? Python's `for` loops are slow, and even aside from that if our matrix is very large then we may be doing more comparisons than is necessary.
 
https://github.com/yohanchatelain/sigdigits/blob/2b4de1d895bab8db8f50ee5ff7979a03109e3941/sigdigits.py#L147
 
* Q: If we have different significance along different features of our samples, won't this overwrite the significance for all of them to be that of the feature with the highest significance?
* Q: Does this work for arbitrary dimensional arrays? If so, it appears as though it expects observations to be stacked in the first axis? Adding a docstring to this effect would be helpful.
"

**Comment by yohanchatelain** [comment]
> This looks great! A few questions that may either be a) entirely irrelevant, or b) a tiny bit helpful  .
> https://github.com/yohanchatelain/sigdigits/blob/2b4de1d895bab8db8f50ee5ff7979a03109e3941/sigdigits.py#L84-L85
> 
> 
> * Q: Do we want to fix a single random seed if we're shuffling, or fix and then compute the value over a few iterations and average them? I don't think we want a measure that changes with random state, if we can avoid it. This question applies a few more times throughout the module.
 
We could let the user to fix a random seed, the paper stated that if you want to compare samples for the same random variable, you need to split your set in two with one for the samples and one for the reference. Normally, the order of your samples shouldn't change the result by much so you can reorder them by shuffling the array. The idea of shuffling is to let the user seeing if there differences due to a "bad" order. But not averaging them since you must compare sample by sample.
 
> https://github.com/yohanchatelain/sigdigits/blob/2b4de1d895bab8db8f50ee5ff7979a03109e3941/sigdigits.py#L142-L149
> 
> 
> * Q: if all we're doing here is comparing the `z` value to the `2^{-k}` power for each iteration, couldn't we just take the `log` of all our values of `z` (that are non-zero) and skip the loop all together? Python's `for` loops are slow, and even aside from that if our matrix is very large then we may be doing more comparisons than is necessary.
 
I understand for the log but I don't see how remove the loop. Since you need to know when you all your values are True to stop.
 
> https://github.com/yohanchatelain/sigdigits/blob/2b4de1d895bab8db8f50ee5ff7979a03109e3941/sigdigits.py#L147
> 
> 
> * Q: If we have different significance along different features of our samples, won't this overwrite the significance for all of them to be that of the feature with the highest significance?
 
Yes you're probably right. It should be like for `contributing_digits_general`.
 
> * Q: Does this work for arbitrary dimensional arrays? If so, it appears as though it expects observations to be stacked in the first axis? Adding a docstring to this effect would be helpful.
 
Yes, I wrote the code by supposing observations to be stacked in the first axis. But I would like to extent it for any axis. Any PR is welcom . And yes a clean documentation needs to be add.

**Comment by gkiar** [comment]
thanks @yohanchatelain ! A few quick replies

- First point (re: random state); yes, I suppose it's a thing worth having a warning raised when the shuffle option is selected, suggesting that they should repeat this process to determine if sorting plays a role.

- I think we could replace the entire loop structure with something like the following. This may need to be a whiteboard discussion if we're still not understanding one another :) these things are tricky over text
  - filter for all locations where z are not zero across all observations (set those to max value)
  - take the log2 of the remaining elements of z (rather than doing a negative power of 2 to get from bits to data space)
  - round up to the nearest integer for the whole matrix (since we do the conversion of all of our data to bits at once, but they're still floats)

- :+1:

- Sounds good :) The code I wrote [here](https://github.com/gkpapers/2020AggregateMCA/blob/master/code/utils.py#L68-L100) works for arbitrary dimensioned data and assumed stacking along the last axis, so we can merge or consolidate logic from the two as it makes sense/is helpful.
